types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+	constraint->pending.cursor_hint.y = wl_fixed_to_double(y);
+	constraint->pending.cursor_hint.valid = true;
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_commit);
+
+	if (constraint->current.region != constraint->pending.region) {
+		send_unconstrained(constraint);
+	}
+	constraint->current = constraint->pending;
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_destroy);
+
+	send_unconstrained(constraint);
+	wl_resource_destroy(constraint->resource); // is this needed?
+}
+
+static const struct zwp_confined_pointer_v1_interface confined_pointer_impl = {
+	.destroy    = resource_destroy,
+	.set_region = pointer_constraint_set_region,
+};
+
+static const struct zwp_locked_pointer_v1_interface locked_pointer_impl = {
+	.destroy                  = resource_destroy,
+	.set_cursor_position_hint = pointer_constraint_set_cursor_position_hint,
+	.set_region               = pointer_constraint_set_region,
+};
+
+static void pointer_constraint_create(struct wl_client *client,
+		struct wl_resource *pointer_constraints_resource, uint32_t id,
+		struct wl_resource *surface_resource, struct wl_resource *pointer,
+		struct wl_resource *region, uint32_t lifetime,
+		enum wlr_pointer_constraint_v1_type type) {
+	struct wlr_pointer_constraints_v1 *pointer_constraints = pointer_constraints_from_resource(pointer_constraints_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
+
+	struct wl_list *list = &surface->pointer_constraints;
+
+	int version = wl_resource_get_version(pointer_constraints_resource);
+
+	bool locked_pointer = type == WLR_POINTER_CONSTRAINT_V1_LOCKED;
+
+	struct wl_resource *resource = locked_pointer ?
+		wl_resource_create(client, &zwp_locked_pointer_v1_interface, version, id) :
+		wl_resource_create(client, &zwp_confined_pointer_v1_interface, version, id);
+
+	if (resource == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_pointer_resource(pointer);
+
+	struct wlr_pointer_constraint_v1 *constraint;
+	wl_list_for_each(constraint, list, link) {
+		if (constraint->seat_client == seat_client) {
+			wl_resource_post_error(resource,
+				ZWP_POINTER_CONSTRAINTS_V1_ERROR_ALREADY_CONSTRAINED,
+				"a pointer constraint with a wl_pointer of the same wl_seat"
+				" is already on this surface");
+			return;
+		}
+	}
+
+	constraint = calloc(1, sizeof(struct wlr_pointer_constraint_v1));
+	if (constraint == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	constraint->resource = resource;
+	constraint->surface = surface;
+	constraint->seat_client = seat_client;
+	constraint->current.region = region;
+	constraint->pending.region = region;
```

Comment:
The protocol says you have to handle a NULL region here by using the input region of the surface as well. Not sure if it's copy semantics or is updated when surface input region changes.

Reply:
I don't think it's copy semantics, because that would be quite inconvenient for clients.